commcare/src/main/java/org/motechproject/commcare/tasks/builder/ReportActionBuilder.java

```diff
@@ -0,0 +1,215 @@
+package org.motechproject.commcare.tasks.builder;
+
+import org.motechproject.commcare.config.Config;
+import org.motechproject.commcare.domain.report.ReportMetadataFilter;
+import org.motechproject.commcare.domain.report.ReportMetadataInfo;
+import org.motechproject.commcare.domain.report.ReportsMetadataInfo;
+import org.motechproject.commcare.domain.report.constants.FilterDataType;
+import org.motechproject.commcare.domain.report.constants.FilterType;
+import org.motechproject.commcare.events.constants.DisplayNames;
+import org.motechproject.commcare.events.constants.EventDataKeys;
+import org.motechproject.commcare.events.constants.EventSubjects;
+import org.motechproject.commcare.service.CommcareConfigService;
+import org.motechproject.commcare.service.CommcareSchemaService;
+import org.motechproject.tasks.contract.ActionEventRequest;
+import org.motechproject.tasks.contract.ActionParameterRequest;
+import org.motechproject.tasks.contract.builder.ActionEventRequestBuilder;
+import org.motechproject.tasks.contract.builder.ActionParameterRequestBuilder;
+import org.motechproject.tasks.domain.enums.ParameterType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * Class responsible for building the Task module actions, related to sending the
+ * forms via submission API. The built list of {@link ActionEventRequest}s instances can
+ * be passed to the Task module to register channel actions.
+ */
+public class ReportActionBuilder implements ActionBuilder {
+
+    private CommcareSchemaService schemaService;
+    private CommcareConfigService configService;
+
+    public ReportActionBuilder (CommcareSchemaService schemaService, CommcareConfigService configService) {
+        this.schemaService = schemaService;
+        this.configService = configService;
+    }
+
+    @Override
+    public List<ActionEventRequest> buildActions () {
+        List<ActionEventRequest> actions = new ArrayList<>();
+
+        for (Config config : configService.getConfigs().getConfigs()) {
+            List<ActionEventRequest> builtActions = buildActionsForConfigs(config);
+            actions.addAll(builtActions);
+        }
+
+        return actions;
+    }
+
+    private List<ActionEventRequest> buildActionsForConfigs (Config config) {
+        List<ActionEventRequest> actions = new ArrayList<>();
+
+        for (ReportsMetadataInfo reportsMetadataInfo : schemaService.getReportsMetadata(config.getName())) {
```

Comment:
method with objects connected with db (entity annotation) should be in transaction (please remember that transactional annotation only works for public methods, so in thi case add this to buildActions method